src/Middleware/Rewrite/src/RedirectToNonWwwRule.cs

```diff
@@ -0,0 +1,60 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Rewrite.Logging;
+
+namespace Microsoft.AspNetCore.Rewrite
+{
+    internal class RedirectToNonWwwRule : IRule
+    {
+        private const string WwwDot = "www.";
+
+        private readonly string[] _domains;
+        private readonly int _statusCode;
+
+        public RedirectToNonWwwRule(int statusCode)
+        {
+            _statusCode = statusCode;
+        }
+
+        public RedirectToNonWwwRule(int statusCode, params string[] domains)
+        {
+            if (domains == null)
+                throw new ArgumentNullException(nameof(domains));
+
+            if (domains.Length < 1)
+                throw new ArgumentException($"One or more {nameof(domains)} must be provided.");
+
+            _domains = domains;
+            _statusCode = statusCode;
+        }
+
+        public void ApplyRule(RewriteContext context)
+        {
+            var request = context.HttpContext.Request;
+
+            var hostInDomains = RedirectToWwwHelper.IsHostInDomains(request, _domains);
+
+            if (!hostInDomains)
+            {
+                context.Result = RuleResult.ContinueRules;
+                return;
+            }
+
+            if (!request.Host.Value.StartsWith(WwwDot, StringComparison.OrdinalIgnoreCase))
```

Comment:
Shouldn't this be checked before checking `IsInHostDomains`? That way, if a request doesn't start with www, we can short circuit the rule. Also, checking IsHostInDomains is fairly expensive.

Reply:
Ah, I see why you did this now, you are now checking localhost in IsInHostDomains.

I'm on the fence on this. I think it's better to have the checks for IsLocalHost and Www/NonWww earlier as those are common cases and can be short-circuited to the next rule. However, it's more likely that _domains will be null, meaning that all it is doing is a null check.